### PR TITLE
Add Build publish to PyPI and create GitHub release workflow

### PR DESCRIPTION
**Pull request by karan6181**
## Description
- Add Build and publish to PyPI workflow
- Add create GitHub release workflow
- Update existing actions to latest version

## Usage
The `pypi-release.yaml` workflow will be kick in with the below commands
```
git tag -a "v0.0.2b" -m "pypi package versioning"
git push --tags
```

With the above 2 lines, the GitHub actions will do the following
- Run pre-commit for linting issue
- Run pytest to ensure it passes
- Create a pypi wheel, check package description on PyPI using `twine`, uploads the pypi wheel to pypi website.
- Create a GitHub Release

### CODE REVIEW

**Review by knighton — APPROVED** [review]

File: .github/workflows/precommit.yaml
Comment (bandish-shah) [inline .github/workflows/precommit.yaml:24]: It looks like the `install` workflow runs on Python 3.7, 3.8 and 3.9.  Why does this only run on 3.9?

File: .github/workflows/pytest.yaml
Comment (bandish-shah) [inline .github/workflows/pytest.yaml:21]: It looks like the `install` workflow runs on Python 3.7, 3.8 and 3.9.  Why does this only run on 3.9?

File: .github/workflows/pypi-release.yaml
Comment (bandish-shah) [inline .github/workflows/pypi-release.yaml:6]: This will run on any tag created on `dev` or `main`.  Is that what you want?

File: .github/workflows/precommit.yaml
Comment (bandish-shah) [inline .github/workflows/precommit.yaml:49]: IIRC you have to do a `pre-commit install` before you can run or does this go ahead and install `pre-commit` as well?

File: .github/workflows/precommit.yaml
Comment (bandish-shah) [inline .github/workflows/precommit.yaml:2]: Should probably just call this workflow "Linting" or "Code Quality" or something.  Yes we're using `pre-commit` but that's more of a runner vs. being descriptive of what checks are actually being performed here.

**Review by bandish-shah — COMMENTED** [review]
This looks great!  Few questions inline before I approve.

File: .github/workflows/precommit.yaml
Comment (karan6181) [reply .github/workflows/precommit.yaml:24]: We want to make sure that the streaming installation works as expected on 3.7, 3.8 and 3.9. However, we are running the pre-commit workflow on the latest python to make sure it passes since it is just used for linting. Making it to run on a multiple python version is just adding more redundancy without having much benefit.

File: .github/workflows/pytest.yaml
Comment (karan6181) [reply .github/workflows/pytest.yaml:21]: Same as above.

File: .github/workflows/precommit.yaml
Comment (karan6181) [reply .github/workflows/precommit.yaml:49]: pre-commit is getting installed as part of [dev](https://github.com/mosaicml/streaming/blob/main/setup.py#L47) in setup.py file and all the dependency linting will get install inside command `pre-commit run --all-files`.

File: .github/workflows/pypi-release.yaml
Comment (bandish-shah) [reply .github/workflows/pypi-release.yaml:6]: Scratch this, I forgot there was only a `main` branch in this repo.

File: .github/workflows/precommit.yaml
Comment (karan6181) [reply .github/workflows/precommit.yaml:2]: Done in https://github.com/mosaicml/streaming/pull/50/commits/052061bd11fb8c286bb726c73bf10fec330e1398.

Comment (bandish-shah) [reply .github/workflows/precommit.yaml:2]: We should rename this file as well to match the workflow name.

File: .github/workflows/precommit.yaml
Comment (bandish-shah) [reply .github/workflows/precommit.yaml:24]: I think pyright has different linting rules per Python version so technically it isn't redundant.  It might be overkill though so let's keep it minimal for now and we can add additional versions if necessary.

**Review by bandish-shah — APPROVED** [review]
Couple of follow ups, approving to unblock.

File: .github/workflows/precommit.yaml
Comment (karan6181) [reply .github/workflows/precommit.yaml:2]: Done in https://github.com/mosaicml/streaming/pull/50/commits/796fbdbbbf7cad3179d8651314f4f7f171d10ac0.

**Pull request merged**
